### ISSUE TITLE
Characterize rvalue proxy references for concept Writable

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#226]
> If E is an xvalue (3.10), the resulting state of the object it denotes is unspecified. [ Note: The object must still meet the requirements of any library component that is using it. The operations listed in those requirements must work as specified whether the object has been moved from or not.—end note ]
> [Editor’s note: “If E is an xvalue” suffices for now, but incorporating P0022’s proxy iterators will require a characterization of the “xvalue-ness” of a proxy reference type.]

So, it's time to integrate P0022 and we still don't have a satisfactory characterization of the xvalue-ness of a proxy reference type. Arg.

**Comment by CaseyCarter** [comment]
I'm researching to get back the mindset where this made sense so I can fix it. Right now, I observe that the concept says nothing about the resulting state of the object denoted by `t` when `t` is an lvalue. There seems to be an implicit assumption that the assignment expression is non-modifying, and this "If `E` is an xvalue" rule is providing an exception to that general case.

**Comment by CaseyCarter** [comment]
Yes, this is attempting (and failing) to characterize when the assignment expression in `Writable` may and may not modify its RHS on the basis of the value category of the RHS subexpression. The old design captured this distinction in the difference between `MoveWritable` (may modify) and `Writable` (must not modify). Here I'm saying that `Writable` may modify its RHS if it is an xvalue (or an "xvalue-like" proxy) and implicitly (which should be explicit) must not modify its RHS otherwise.

Formulating `Writable` in terms of the output iterator and the expression to be written is great in that it enables the single concept to constrain `*o = *i`, `*o = iter_move(*i)`, `*o = t`, and `*o = std::move(t)` (even when `*i` is an xvalue). It is, however, a PITA in that it requires us to characterize the behavior of the expression purely via the type and category of the RHS. (Much as is the case for `Assignable`.)

**Comment by CaseyCarter** [comment]
Strawman:

1 The `Writable` concept specifies the requirements for writing a value into an iterator’s referenced object.

```c++
template <class>
constexpr bool is_xvalue_like = false;
template <class T>
constexpr bool is_xvalue_like<T&&> = true;

template <class Out, class T>
concept bool Writable() {
  return Semiregular<Out>() &&
    requires(Out o, T&& t) {
      *o = std::forward<T>(t); // not required to be equality preserving
    };
}
```

2 Let `E` be an an expression such that `decltype(E)` is `T`, and let `o` be a dereferenceable object of type `Out`. Then `Writable<Out, T>()` is satisfied if and only if
(2.1) — If `Readable<Out>() && Same<value_type_t<Out>, decay_t<T>>()` is satisfied, then `*o` after the assignment is equal to the value of `E` before the assignment.
(2.2) — if `is_xvalue_like<T>` is `false`, the assignment expression does not modify `t`.

  [...]

**Comment by ericniebler** [comment]
I really dislike the `is_xvalue_like` trait.

Here's my own strawman: we leave `Writable` alone. It can only modify `t` if it really is a non-`const` xvalue. That's perfectly fine for `IndirectlyMovable` with regular or proxy iterators, and for `IndirectlyCopyable` with regular iterators. It's only a problem for `IndirectlyCopyable` with proxy iterators, since (assuming `*i` is an rvalue proxy) it permits `*o = *i` to mutate the value of `*i`. So, we fix it up in `IndirectlyCopyable` by adding an additional semantic constraint that the value of `*i` is not mutated by `*o = *i`. (I've looked at the other uses of `Writable` and `OutputIterator` and they don't seem to be affected by this proxy iterator business.)

Counter-example: a _proxy_ `move_iterator` for which `*i` is actually an prvalue proxy for an xvalue reference. In _that_ case, we actually _do_ want `*o = *i` to be able to mutate the value of `*i`. My head hurts. Need to rest my brain.

**Comment by CaseyCarter** [comment]
> I really dislike the `is_xvalue_like` trait.

As do I. The silly 'disable_foo` traits are obscure ways to use broken types with the library; I could not care less that they are awkward to use. This trait is right in the user's face telling them that we haven't yet figured out the proper way to make this corner of the design work.

> So, we fix it up in `IndirectlyCopyable` by adding an additional semantic constraint that the value of `*i` is not mutated by `*o = *i`.

Which, as you've noted, breaks xvalue-ish input iterators which was the only problem case to begin with.

**Comment by ericniebler** [comment]
(Copied from #299)

Characterizing reference-like types:

Strawman Approach 1:
----
"A _techterm{reference-like}_ type is a class type such that objects of that type have values that are comprised, in whole or in part, from parts outside the object's storage [or a reference to such?]."

Strawman Approach 2:
----
"A _techterm{reference-like}_ type is a class type whose copy constructor is not _equality preserving_ because (copies do not have independent values|values may not be stable) [, or a reference to such class type?]."

Or something. We may also need to say something about an object's "value", since it doesn't seem to be characterized anywhere. Loosely in this context it should mean any property of the object that may be observed, either directly or indirectly.

We would want to give examples of reference-like types, such as `reference_wrapper`, the `tuple` returned by `std::tie`, and `vector<bool>::reference`. 

OK @CaseyCarter , rip it apart.

**Comment by ericniebler** [comment]
As for converting a reference to a value, we could make a customization point for that, perhaps called `decay_copy`, which is either the perfect name or the worst one, I don't know which. Obviously a similar metabolism can be made from the CPO with `decltype`.

**Comment by ericniebler** [comment]
Handwave:
> If `E` is an xvalue <ins>or the result of calling `iter_move`</ins>

?
